File: Reactor/ex2/src/main/java/utils/FlightFactory.java
package utils;

import datamodels.Flight;
import datamodels.FlightRequest;
import lombok.Builder;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

/**
 * This class creates a list random round flight flights for any number
 * of airlines, airports, travel dates, and number of daily flights.
 */
public class FlightFactory {
    private static final Random random = new Random();
    private static final double MIN_USD_PRICE_PER_KM = 0.1;
    private static final double MAX_USD_PRICE_PER_KM = 0.3;
    private static final int BASE_PRICE = 50;
    private static final int MIN_DISTANCE_KM = 200;
    private static final int MAX_DISTANCE_KM = 5000;
    private static final int FLIGHT_SPEED_KPH = 400;
    private static final int RUNWAY_MINUTES = 15;

    private static final int AIRPORTS = 3;
    private static final int AIRLINES = 2;
    private static final int AIRLINE_DAILY_FLIGHTS = 1;
    private static final int DAYS = 3;
    private static final String CURRENCY = "USD";

    @Builder
    private static class FlightPath {
        String fromAirport;
        String toAirport;
        int distance;
    }

    /**
     * Builds a list of flights matching a flight request built using the
     * first flight in a random list of flights. This method should only
     * be used for FLIGHT and BEST_PRICE requests.
     *
     * @return A random list of Flight objects.
     */
    public static List<Flight> buildExpectedFlights(String currency, List<String> currencies) {
        List<Flight> flights = FlightFactory.builder().build();
        FlightRequest flightRequest = FlightFactory.buildRequestFrom(flights.get(0));
        flightRequest.setCurrency(currency);

        return flights.stream()
                .filter(flight ->
                        flight.getDepartureAirport()
                                .equals(flightRequest.getDepartureAirport()) &&
                                flight.getDepartureDate()
                                        .equals(flightRequest.getDepartureDate()) &&
                                flight.getArrivalAirport()
                                        .equals(flightRequest.getArrivalAirport()))
                .peek(flight -> {
                    String randomCurrency = currencies.get(random.nextInt(currencies.size()));
                    flight.setCurrency(randomCurrency);
                })
                .collect(Collectors.toList());
    }

    public static RandomFlightBuilder builder() {
        return new RandomFlightBuilder();
    }

    public static Flight randomFlight() {
        return builder()
                .airlines(1)
                .airports(2)
                .dailyFlights(1)
                .build().get(0);
    }

    public static class RandomFlightBuilder {
        int dailyFlights = AIRLINE_DAILY_FLIGHTS;
        LocalDate fromDate = LocalDate.now();
        LocalDate toDate = fromDate.plusDays(DAYS);
        String currency = CURRENCY;
        List<String> airports = randomStrings(AIRPORTS);
        List<String> airlines = randomStrings(AIRLINES);
        int minDistanceKm = MIN_DISTANCE_KM;
        int maxDistanceKm = MAX_DISTANCE_KM;
        double minPricePerKm = MIN_USD_PRICE_PER_KM;
        double maxPricePerKm = MAX_USD_PRICE_PER_KM;
        int basePrice = BASE_PRICE;

        public RandomFlightBuilder basePrice(int basePrice) {
            this.basePrice = basePrice;
            return this;
        }

        public RandomFlightBuilder airlines(int airlines) {
            this.airlines = randomStrings(airlines);
            return this;
        }

        public RandomFlightBuilder airlines(String... airlines) {
            airlines(Arrays.asList(airlines));
            return this;
        }

        public RandomFlightBuilder airlines(List<String> airlines) {
            this.airlines = airlines.stream()
                    .distinct()
                    .collect(Collectors.toList());
            return this;
        }

        public RandomFlightBuilder airports(int airports) {
            this.airports = randomStrings(airports);
            return this;
        }

        public RandomFlightBuilder airports(String... airports) {
            airlines(Arrays.asList(airports));
            return this;
        }

        public RandomFlightBuilder airports(List<String> airports) {
            this.airports = airports.stream()
                    .distinct()
                    .collect(Collectors.toList());
            return this;
        }

        public RandomFlightBuilder dailyFlights(int dailyFlights) {
            this.dailyFlights = dailyFlights;
            return this;
        }

        public RandomFlightBuilder minDistanceKm(int minDistance) {
            this.minDistanceKm = minDistance;
            return this;
        }

        public RandomFlightBuilder maxDistanceKm(int maxDistance) {
            this.maxDistanceKm = maxDistance;
            return this;
        }

        public RandomFlightBuilder minPricePerKm(double minPricePerKm) {
            this.minPricePerKm = minPricePerKm;
            return this;
        }

        public RandomFlightBuilder maxPricePerKm(double maxPricePerKm) {
            this.maxPricePerKm = maxPricePerKm;
            return this;
        }

        public RandomFlightBuilder from(LocalDate fromDate) {
            this.fromDate = fromDate;
            return this;
        }

        public RandomFlightBuilder from(String fromDate) {
            this.fromDate = LocalDate.parse(fromDate);
            return this;
        }

        public RandomFlightBuilder to(LocalDate toDate) {
            this.toDate = toDate;
            return this;
        }

        public RandomFlightBuilder to(String toDate) {
            this.toDate = LocalDate.parse(toDate);
            return this;
        }

        public RandomFlightBuilder currency(String currency) {
            this.currency = currency;
            return this;
        }

        public List<Flight> build() {
            return randomFlights(this);
        }
    }

    private static List<Flight> randomFlights(RandomFlightBuilder builder) {
        List<FlightPath> flightPaths = new ArrayList<>();
        for (int i = 0; i < builder.airports.size() - 1; i++) {
            String fromAirport = builder.airports.get(i);
            for (int j = i + 1; j < builder.airports.size(); j++) {
                String toAirport = builder.airports.get(j);
                int distance = random(builder.minDistanceKm, builder.maxDistanceKm);
                flightPaths.add(FlightPath.builder()
                        .fromAirport(fromAirport)
                        .toAirport(toAirport)
                        .distance(distance)
                        .build());
                flightPaths.add(FlightPath.builder()
                        .fromAirport(toAirport)
                        .toAirport(fromAirport)
                        .distance(distance)
                        .build());
            }
        }

        List<Flight> flights =
                builder.fromDate.datesUntil(builder.toDate.plusDays(1)).flatMap(date ->
                        builder.airlines.stream().flatMap(airline ->
                                flightPaths.stream().flatMap(flight ->
                                        IntStream.range(0, builder.dailyFlights).mapToObj(__ ->
                                                randomFlight(
                                                        flight.fromAirport,
                                                        flight.toAirport,
                                                        date,
                                                        flight.distance,
                                                        builder.basePrice,
                                                        builder.minPricePerKm,
                                                        builder.maxPricePerKm,
                                                        airline,
                                                        builder.currency)
                                        )
                                )
                        )
                ).collect(Collectors.toList());

        long days = builder.fromDate.datesUntil(builder.toDate.plusDays(1)).count();
        long expected = builder.airlines.size()
                * flightPaths.size()
                * days
                * builder.dailyFlights;

        return flights;
    }

    private static Flight randomFlight(
            String departureAirport,
            String arrivalAirport,
            LocalDate date,
            int distance,
            int basePrice,
            double minPricePerKm,
            double maxPricePerKm,
            String airlineCode,
            String currency
    ) {

        LocalTime flightTime = flightTime(distance);
        LocalTime departureTime = randomDepartureTime(flightTime);
        LocalTime arrivalTime = arrivalTime(departureTime, flightTime);

        return Flight.builder()
                .departureAirport(departureAirport)
                .departureDate(date)
                .departureTime(departureTime)
                .arrivalAirport(arrivalAirport)
                .arrivalTime(arrivalTime)
                .arrivalDate(date)
                .airlineCode(airlineCode)
                .kilometers(distance)
                .price(randomPrice(distance, basePrice, minPricePerKm, maxPricePerKm))
                .currency(currency)
                .build();
    }

    public static FlightRequest buildRequestFrom(Flight flight) {
        return FlightRequest.builder()
                .departureAirport(flight.getDepartureAirport())
                .departureDate(flight.getDepartureDate())
                .arrivalAirport(flight.getArrivalAirport())
                .currency(flight.getCurrency())
                .build();
    }

    private static double random(double min, double max) {
        return min + random.nextDouble() * (max - min);
    }

    private static int random(int min, int max) {
        return (int) random((double) min, max);
    }

    private static int randomPrice(
            int distance, int basePrice, double min, double max) {
        return basePrice + (int) (distance * random(min, max));
    }

    private static LocalTime flightTime(int distance) {
        return LocalTime.ofSecondOfDay(RUNWAY_MINUTES +
                (long) ((double) distance / FLIGHT_SPEED_KPH * 60 * 60));
    }

    private static LocalTime randomDepartureTime(LocalTime flightTime) {
        // Latest departure hour 24 hours - flightTime.hours.
        return LocalTime.of(
                random.nextInt(23 - flightTime.getHour()),
                random.nextInt(60));
    }

    private static LocalTime arrivalTime(LocalTime departureTime, LocalTime flightTime) {
        return LocalTime.from(departureTime)
                .plusHours(flightTime.getHour())
                .plusMinutes(flightTime.getMinute());
    }

    private static List<String> randomStrings(int count) {
        List<String> strings = new ArrayList<>();
        int length = (int) (Math.log10(count) + 1);
        for (int i = 1; i <= count; i++) {
            strings.add(String.format("%0" + length + "d", i));
        }

        return strings;
    }

    /**
     * For testing.
     */
    public static void main(String[] args) {
        List<Flight> flights = builder().airlines("AA").build();

        flights.sort(Comparator.comparing(Flight::getDepartureDate));
        System.out.println(Flight.toSqlInsertString(flights));
    }
}

File: Reactor/ex2/src/main/java/utils/FlightFactory.kt
package utils

import datamodels.Flight
import datamodels.FlightRequest
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.util.*
import java.util.stream.Collectors
import kotlin.math.log10
import kotlin.math.min

/**
 * This class creates a list random round flight flights for any
 * number of airlines, airports, travel dates, and number of daily
 * flights.
 */
object FlightFactory {
    private const val AIRLINES = 5
    private const val DAYS = 1
    private const val BASE_PRICE_USD = 100
    private const val MIN_PRICE_PER_KM_USD = 0.4
    private const val MAX_PRICE_PER_KM_USD = 0.7
    private const val MIN_DISTANCE_KM = 200
    private const val MAX_DISTANCE_KM = 5000
    private const val FLIGHT_SPEED_KPH = 400
    private const val RUNWAY_MINUTES = 15
    private const val DAILY_FLIGHTS = 100
    private const val MIN_LOWEST_PRICE_MATCHES = 1
    private const val MAX_LOWEST_PRICE_MATCHES = 3

    private val random = Random()

    private val expectedMatches = mutableListOf<Flight>()

    @JvmStatic
    fun buildFlightRequestFrom(flights: List<Flight>): FlightRequest {
        return flights[random(0, flights.size)].buildRequest()
    }

    class Builder {
        var airports = randomStrings(2, prefix = "Airport-")
        var airlines = randomStrings(AIRLINES, prefix = "Airline-")
        var fromAirport = airports.first()
        var toAirport = airports.last()
        var minFlights = DAILY_FLIGHTS
        var maxFlights = DAILY_FLIGHTS
        var fromDate: LocalDate = LocalDate.now()
        var toDate: LocalDate = fromDate.plusDays(DAYS.toLong())
        var currencies = ExchangeRate.Currency.values().map { it.toString() }
        var minDistanceKm = MIN_DISTANCE_KM
        var maxDistanceKm = MAX_DISTANCE_KM
        var minPricePerKm = MIN_PRICE_PER_KM_USD
        var maxPricePerKm = MAX_PRICE_PER_KM_USD
        var basePrice = BASE_PRICE_USD
        var minLowestPriceMatches = MIN_LOWEST_PRICE_MATCHES
        var maxLowestPriceMatches = MAX_LOWEST_PRICE_MATCHES

        fun validate() {
            check(minLowestPriceMatches in 1..maxLowestPriceMatches)
            check(minFlights in 1..maxFlights)
            check(minDistanceKm in 1..maxDistanceKm)
            check(0.0 < minPricePerKm && minPricePerKm <= maxPricePerKm)
            check(airports.isNotEmpty())
            check(airlines.isNotEmpty())
            check(currencies.isNotEmpty())
            check(fromDate <= toDate)
            check(minFlights - maxLowestPriceMatches > 0)
        }

        fun minDailyFlights(min: Int): Builder {
            this.minFlights = min
            return this
        }

        fun maxDailyFlights(max: Int): Builder {
            this.maxFlights = max
            return this
        }

        fun minLowestPriceMatches(matches: Int): Builder {
            this.minLowestPriceMatches = matches
            return this
        }

        fun maxLowestPriceMatches(matches: Int): Builder {
            this.maxLowestPriceMatches = matches
            return this
        }

        fun basePrice(basePrice: Int): Builder {
            this.basePrice = basePrice
            return this
        }

        fun airlines(airlines: Int): Builder {
            airlines(randomStrings(airlines))
            return this
        }

        fun airlines(vararg airlines: String): Builder {
            airlines(listOf(*airlines))
            return this
        }

        fun airlines(airlines: List<String>): Builder {
            this.airlines = airlines.stream()
                .distinct()
                .collect(Collectors.toList())
            return this
        }

        fun fromAirport(airport: String): Builder {
            fromAirport = airport
            return this
        }

        fun toAirport(airport: String): Builder {
            toAirport = airport
            return this
        }

        fun minFlights(min: Int): Builder {
            this.minFlights = min
            return this
        }

        fun maxFlights(max: Int): Builder {
            this.maxFlights = max
            return this
        }

        fun minDistanceKm(minDistance: Int): Builder {
            minDistanceKm = minDistance
            return this
        }

        fun maxDistanceKm(maxDistance: Int): Builder {
            maxDistanceKm = maxDistance
            return this
        }

        fun minPricePerKm(minPricePerKm: Double): Builder {
            this.minPricePerKm = minPricePerKm
            return this
        }

        fun maxPricePerKm(maxPricePerKm: Double): Builder {
            this.maxPricePerKm = maxPricePerKm
            return this
        }

        fun from(fromDate: LocalDate): Builder {
            this.fromDate = fromDate
            return this
        }

        fun from(fromDate: String?): Builder {
            this.fromDate = LocalDate.parse(fromDate)
            return this
        }

        fun to(toDate: LocalDate): Builder {
            this.toDate = toDate
            return this
        }

        fun to(toDate: String?): Builder {
            this.toDate = LocalDate.parse(toDate)
            return this
        }

        fun currencies(currencies: List<String>): Builder =
            this.also {
                it.currencies = currencies
            }

        fun generateFlights(): List<Flight> {
            validate()

            expectedMatches.clear()

            var minPrice = Double.MAX_VALUE
            var flights = mutableListOf<Flight>()

            val distance = random(minDistanceKm, maxDistanceKm + 1)

            var flightCount = random(minFlights, maxFlights + 1)
            val matches = random(minLowestPriceMatches, maxLowestPriceMatches + 1)

            (1..flightCount).map { count ->
                val flightTime = flightTime(distance)
                val departureDateTime =
                    LocalDateTime.of(fromDate, randomDepartureTime(flightTime))
                val arrivalDateTime =
                    departureDateTime
                        .plusHours(flightTime.hour.toLong())
                        .plusMinutes(flightTime.minute.toLong())


                val price =
                    if (count <= flightCount - matches) {
                        randomPrice(
                            distance,
                            basePrice,
                            minPricePerKm,
                            maxPricePerKm
                        ).toDouble()
                            .also {
                                minPrice = min(minPrice, it)
                            }
                    } else {
                        minPrice - 10.0
                    }

                val currency = currencies.random()

                val flight = Flight(
                    count.toLong(),
                    fromAirport,
                    departureDateTime.toLocalDate(),
                    departureDateTime.toLocalTime(),
                    toAirport,
                    arrivalDateTime.toLocalDate(),
                    arrivalDateTime.toLocalTime(),
                    distance,
                    ExchangeRate.convert(price, "USD", currency),
                    currency,
                    airlines.random()
                )

                flights.add(flight)
            }

            check(flights.count() in minFlights..maxFlights)

            val shuffled = flights.shuffled(random)

            val min = shuffled.minOf { ExchangeRate.convert(it.price, it.currency, "USD") }
            val cheapest = shuffled.filter { ExchangeRate.convert(it.price, it.currency, "USD") == min }

            println("FlightFactory has generated ${shuffled.count()} " +
                    "flights with ${cheapest.count()} cheapest flight${if (cheapest.count() > 1) "s" else ""}.")
            println("\nCONFIGURATION:\n")
            printConfiguration()

            println("\nCHEAPEST FLIGHT${if (cheapest.count() > 1) "S" else ""}:\n")
            cheapest.forEach {
                println("\t$it")
            }
            println()

            check(cheapest.count() in minLowestPriceMatches..maxLowestPriceMatches)


            return shuffled
        }

        private fun printConfiguration() {
            println(
                "\tfromAirport = $fromAirport\n" +
                        "\ttoAirport = $toAirport\n" +
                        "\tminFlights = $minFlights\n" +
                        "\tmaxFlights = $maxFlights\n" +
                        "\tfromDate = $fromDate\n" +
                        "\ttoDate = $toDate\n" +
                        "\tcurrencies = $currencies\n" +
                        "\tairlines = $airlines\n" +
                        "\tminDistanceKm = $minDistanceKm\n" +
                        "\tmaxDistanceKm = $maxDistanceKm\n" +
                        "\tminPricePerKm = $minPricePerKm\n" +
                        "\tmaxPricePerKm = $maxPricePerKm\n" +
                        "\tbasePrice = $basePrice\n" +
                        "\tminLowestPriceMatches = $minLowestPriceMatches\n" +
                        "\tmaxLowestPriceMatches = $maxLowestPriceMatches"
            )
        }
    }

    private fun randomStrings(count: Int, prefix: String = ""): List<String> {
        val strings: MutableList<String> = ArrayList()
        val length = (log10(count.toDouble()) + 1).toInt()
        for (i in 1..count) {
            strings.add(String.format("$prefix%0" + length + "d", i))
        }
        return strings
    }

    private fun random(min: Double, max: Double): Double {
        return min + random.nextDouble() * (max - min)
    }

    private fun random(min: Int, max: Int): Int {
        return random(min.toDouble(), max.toDouble()).toInt()
    }

    private fun randomPrice(
        distance: Int, basePrice: Int, min: Double, max: Double
    ): Int {
        return basePrice + (distance * random(min, max)).toInt()
    }

    private fun flightTime(distance: Int): LocalTime {
        return LocalTime.ofSecondOfDay(
            RUNWAY_MINUTES + (distance.toDouble() / FLIGHT_SPEED_KPH * 60 * 60).toLong()
        )
    }

    /**
     * Produces a random departure time that ensures that the
     * arrival time will be before midnight on the same day.
     */
    private fun randomDepartureTime(flightTime: LocalTime): LocalTime {
        // Latest departure hour 24 hours - flightTime.hours.
        check(flightTime.hour <= 23)
        random.nextInt(flightTime.minute)

        val hour: Int
        val min: Int

        return try {
            hour = random.nextInt(23 - flightTime.hour)
            min = random.nextInt(flightTime.minute)

            LocalTime.of(hour, min)
        } catch (e: Exception) {
            throw e
        }
    }

    private fun arrivalTime(departureTime: LocalTime, flightTime: LocalTime): LocalTime =
        LocalTime.from(departureTime)
            .plusHours(flightTime.hour.toLong())
            .plusMinutes(flightTime.minute.toLong())
}
